zypp-core/zyppng/base/zyppglobal.h

```diff
@@ -36,10 +36,12 @@ template <typename Ptr> inline auto zyppGetPtrHelper(Ptr &ptr) -> decltype(ptr.o
     const Class##Private* Class::d_func() const \
     { return reinterpret_cast<const Class##Private *>(zyppGetPtrHelper(d_ptr)); }
 
-#define ZYPP_DECLARE_PUBLIC(Class)                                    \
```

Comment:
L14: `those macros are inspired by the Qt framework`: Please leave a few more words about why, how and where to use them.

I'm a bit concerned about the `reinterpret_cast` because it won't work in cases of multiple inheritance, though 
statemachine.h:345  `* class MyStateMachine : public SmBase<MyStateMachine>, public Base { };` suggests using it. 
If the public class uses the macros the `*Private` class and its `BasePrivate` must share the same address, that's not always true. In this case a caveat shoul be added here and in the docs of `BasePrivate`.

For bug hunting I'd prefer to have a define here that switches all relevant locations to use dynamic_casts (so one could add e.g. a tracer to a Private class)

Reply:
Do I understand your concern correctly:
The *Private and BasePrivate pimpls are always the same because the leaf class always creates the pimpl for ALL types in the inheritance tree.  
Only way this could break is if we use multiple inheritance where more than one base class derives from Base ... but we do not support that. 
The statemachine example btw is not problematic because SmBase is not derived from Base, thus the Statemachine part does not have a pimpl and will not use the macros.